### PR TITLE
boards: arm: Add Renesas Spider board

### PR DESCRIPTION
**Pull request by aaillet**
This pull request is introducing Generation 4 of Renesas R-Car SoCs with the S4 SoC as well as the Spider board into Zephyr.

Gen4 SoCs vs Gen3 (Zephyr POV) :
- Cortex-R52 is now used as Real Time core
- We are now able to use embedded ARM timer
- Some IPs are the same and will use existing drivers, sometime with few improvements (UART, GPIO...)
- Introduction of new IPs

S4 SoC that is equipping multiple Gen4 boards have some Cortex-A55s and a Cortex-R52 core.

Spider board is the reference board for S4 SoC, developpment oriented, this board is providing a lot of available IOs.

Operation of basic samples such as hello_world and blinky have been validated.

### CODE REVIEW

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include <zephyr/dt-bindings/gpio/gpio.h>
+#include <zephyr/dt-bindings/clock/r8a779f0_cpg_mssr.h>
+
+/ {
+	soc {
+		pfc: pin-controller@e6050000 {
+			compatible = "renesas,rcar-pfc";
+			reg = <0xe6050000 0x16c>, <0xe6050800 0x16c>,
+				  <0xe6051000 0x16c>, <0xe6051800 0x16c>,
+				  <0xffd90000 0x16c>, <0xffd90800 0x16c>,
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/r8a779f0.dtsi]: il manque un patch ici, les adresses sont censé être 0xdfd90000 ... pas 0xf)

Comment (aaillet) [reply dts/arm/renesas/rcar/gen4/r8a779f0.dtsi]: fixed !

File: drivers/pinctrl/pfc_rcar.c [hunk 2]
```diff
@@ -10,13 +10,23 @@
...
+/* Gen3 only has one base address, Gen4 has one per GPIO controller */
+#if defined(CONFIG_SOC_SERIES_RCAR_GEN3)
+static const uint32_t reg_base[] = {DT_INST_REG_ADDR(0)};
+#elif defined(CONFIG_SOC_SERIES_RCAR_GEN4)
+/* swap both arguments */
+#define PFC_REG_ADDRESS(idx, node_id) DT_REG_ADDR_BY_IDX(node_id, idx)
+static const uint32_t reg_base[] = {
+	LISTIFY(DT_NUM_REGS(DT_DRV_INST(0)), PFC_REG_ADDRESS, (,), DT_DRV_INST(0))
+};
+
```
Comment (pmarzin) [inline drivers/pinctrl/pfc_rcar.c]: Is it a good idea (or mandatory ?) to add a #else with #error "Not supported" ?

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+			compatible = "renesas,rcar-gpio";
+			reg = <0xe6050180 0x54>;
+			#gpio-cells = <2>;
+			gpio-controller;
+			interrupts = <GIC_SPI 822 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+			clocks = <&cpg CPG_MOD 915>;
+			status = "disabled";
+		};
+
+		/* Cant manage to use GPIOs from PFSS domain buses yet */
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/r8a779f0.dtsi]: should we change the msg here has we know now how to use them ?

File: dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+				<GIC_PPI 10 IRQ_TYPE_LEVEL
+				 IRQ_DEFAULT_PRIORITY>;
+	};
+
+	soc {
+		interrupt-parent = <&gic>;
+
+		sram0: memory@40040000 {
+			compatible = "mmio-sram";
+			reg = <0x40040000 0x1fc0000>;
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi]: not sure about the 0x1fc0000; need to check that size.

Comment (pmarzin) [reply dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi]: <html><body>

H'00_EB200000 | ～ | H'00_EB2FFFFF | RT-SRAM0 | RAM
-- | -- | -- | -- | --

</body></html>

So size should be 0x100000 (=1MB) :) To set the same size of memory as the one really used with RT-SRAM.

File: boards/arm/rcar_spider/Kconfig.board [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline boards/arm/rcar_spider/Kconfig.board]: 2023

File: boards/arm/rcar_spider/Kconfig.defconfig [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline boards/arm/rcar_spider/Kconfig.defconfig]: 2023

File: boards/arm/rcar_spider/rcar_spider_cr52-pinctrl.dtsi [hunk 7]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline boards/arm/rcar_spider/rcar_spider_cr52-pinctrl.dtsi]: 2023

File: boards/arm/rcar_spider/rcar_spider_cr52.dts [hunk 8]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline boards/arm/rcar_spider/rcar_spider_cr52.dts]: 2023

File: soc/arm/renesas_rcar/gen4/Kconfig.defconfig.r8a779f0 [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2021 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/Kconfig.defconfig.r8a779f0]: 2023

File: soc/arm/renesas_rcar/gen4/Kconfig.defconfig.series [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# Renesas R-Car Gen4 SoC line
+
+# Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/Kconfig.defconfig.series]: 2023

File: soc/arm/renesas_rcar/gen4/Kconfig.series [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/Kconfig.series]: 2023

File: soc/arm/renesas_rcar/gen4/Kconfig.soc [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/Kconfig.soc]: 2023

File: soc/arm/renesas_rcar/gen4/linker.ld [hunk 13]
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2021 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/linker.ld]: 2023

File: soc/arm/renesas_rcar/gen4/soc.c [hunk 14]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/soc.c]: 2023

File: soc/arm/renesas_rcar/gen4/soc.h [hunk 15]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/soc.h]: 2023

File: drivers/clock_control/clock_control_renesas_cpg_mssr.h [hunk 16]
```diff
@@ -41,6 +41,37 @@ static const uint16_t srcr[] = {
...
+	0x2C00, 0x2C04, 0x2C08, 0x2C0C,
+	0x2C10, 0x2C14, 0x2C18, 0x2C1C,
+	0x2C20, 0x2C24, 0x2C28, 0x2C2C,
+	0x2C30, 0x2C34, 0x2C38, 0x2C3C,
+	0x2C40, 0x2C44, 0x2C48, 0x2C4C,
+	0x2C50, 0x2C54, 0x2C58, 0x2C5C,
+	0x2C60, 0x2C64, 0x2C68, 0x2C6C,
+};
+
+/* SCIF clock */
```
Comment (pmarzin) [inline drivers/clock_control/clock_control_renesas_cpg_mssr.h]: Use same comment as GEN3 [ligne 43](https://github.com/iotbzh/zephyr/pull/43/commits/294e168bc43bfca8f303eef515062ca914060bcf#diff-a3879d78cc1f7eba632c4da477dc497eeca5ba01899f769c8c7413ec4bebcbe0R43)
```c
/* Peripherals Clocks */
#define S0D12_PER_CLK_RATE        66600000        /* SCIF */
```

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi [hunk 17]
```diff
@@ -1,14 +1,25 @@
...
  */
 
 #include <renesas/rcar/gen4/rcar_gen4_cr52.dtsi>
 #include <zephyr/dt-bindings/gpio/gpio.h>
+#include <zephyr/dt-bindings/clock/r8a779f0_cpg_mssr.h>
 
 / {
 	soc {
+		/* Clock controller
+		 * Using domain 0 as linux
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/r8a779f0.dtsi]: Same as previous commit

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi [hunk 18]
```diff
@@ -1,14 +1,25 @@
...
+		 * Tried domain 3 unsuccessfully
+		 */
+		cpg: clock-controller@e6150000 {
+			compatible = "renesas,r8a779f0-cpg-mssr";
+			reg = <0xe6150000 0x4000>;
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:30]: Same as previous commit

File: dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+			compatible = "arm,gic";
+			reg = <0xf0000000 0x1000>,
+				  <0xf0100000 0x20000>;
+			interrupt-controller;
+			#interrupt-cells = <4>;
+			status = "okay";
+		};
+
+		/* Clock controller
+		 * Using domain 0 as linux
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi]: What is explained in this comment ? Why talking about domain into clock part ?

File: dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+		 * Should use domain 3 later
+		 */
+		cpg: clock-controller@e6150000 {
+			compatible = "renesas,rcar-cpg-mssr";
+			reg = <0xe6150000 0x4000>;
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi]: `0x4000` is same as linux ? Looks to end at `0x3c00` in docs (without clock monitoring)

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi [hunk 21]
```diff
@@ -10,6 +10,14 @@
 
 / {
 	soc {
+		pfc: pin-controller@e6050000 {
+			compatible = "renesas,rcar-pfc";
+			reg = <0xe6050000 0x16c>, <0xe6050800 0x16c>,
+				  <0xe6051000 0x16c>, <0xe6051800 0x16c>,
+				  <0xdfd90000 0x16c>, <0xdfd90800 0x16c>,
```
Comment (pmarzin) [inline dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:21]: Add a comment for address conversion over control domain ?

File: soc/arm/renesas_rcar/gen4/pfc_r8a779f0.c [hunk 22]
```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright (c) 2022-2023 IoT.bzh
```
Comment (pmarzin) [inline soc/arm/renesas_rcar/gen4/pfc_r8a779f0.c]: 2023

File: boards/arm/rcar_spider/rcar_spider_cr52.dts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+	pinctrl-0 = <&scif0_data_tx_default &scif0_data_rx_default>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+/* Linux console */
+&scif3 {
+	pinctrl-0 = <&scif3_data_tx_default &scif3_data_rx_default>;
+	pinctrl-names = "default";
+	status = "okay";
```
Comment (pmarzin) [inline boards/arm/rcar_spider/rcar_spider_cr52.dts]: Should be `disabled` to let Linux alone ?

File: boards/arm/rcar_spider/rcar_spider_cr52_defconfig [hunk 24]
```diff
@@ -0,0 +1,15 @@
+CONFIG_SOC_R8A779F0=y
+CONFIG_SOC_SERIES_RCAR_GEN4=y
+CONFIG_BOARD_RCAR_SPIDER_CR52=y
+CONFIG_CLOCK_CONTROL=y
+#FIXME: What is the correct counter frequency ?
```
Comment (pmarzin) [inline boards/arm/rcar_spider/rcar_spider_cr52_defconfig]: Should be fixed before MERGE

File: boards/arm/rcar_spider/doc/img/rcar_s4_block_diagram.jpg
Comment (pmarzin) [inline boards/arm/rcar_spider/doc/img/rcar_s4_block_diagram.jpg:1]: For me to know, is Zephyr github adding this type of file as LFS ?

**Review by pmarzin — COMMENTED** [review]
Nice job :)

File: boards/arm/rcar_spider/doc/img/rcar_s4_block_diagram.jpg
Comment (aaillet) [reply boards/arm/rcar_spider/doc/img/rcar_s4_block_diagram.jpg:1]: As file, see some work to reduce image sizes some time ago here : https://github.com/zephyrproject-rtos/zephyr/pull/49024

File: drivers/pinctrl/pfc_rcar.c (hunk 2 above)
Comment (aaillet) [reply drivers/pinctrl/pfc_rcar.c]: I think that as we have full control on the code tree , there is no way that this driver can be called by another "SOC SERIE" board.
But even if not used, a protection can be added !

File: boards/arm/rcar_spider/Kconfig.defconfig (hunk 6 above)
Comment (aaillet) [reply boards/arm/rcar_spider/Kconfig.defconfig]: fixed

File: boards/arm/rcar_spider/rcar_spider_cr52-pinctrl.dtsi (hunk 7 above)
Comment (aaillet) [reply boards/arm/rcar_spider/rcar_spider_cr52-pinctrl.dtsi]: fixed

File: boards/arm/rcar_spider/rcar_spider_cr52.dts (hunk 8 above)
Comment (aaillet) [reply boards/arm/rcar_spider/rcar_spider_cr52.dts]: fixed

File: soc/arm/renesas_rcar/gen4/Kconfig.defconfig.r8a779f0 (hunk 9 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/Kconfig.defconfig.r8a779f0]: fixed

File: soc/arm/renesas_rcar/gen4/Kconfig.defconfig.series (hunk 10 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/Kconfig.defconfig.series]: fixed

File: soc/arm/renesas_rcar/gen4/Kconfig.series (hunk 11 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/Kconfig.series]: fixed

File: soc/arm/renesas_rcar/gen4/Kconfig.soc (hunk 12 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/Kconfig.soc]: fixed

File: soc/arm/renesas_rcar/gen4/linker.ld (hunk 13 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/linker.ld]: fixed

File: soc/arm/renesas_rcar/gen4/soc.c (hunk 14 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/soc.c]: fixed

File: soc/arm/renesas_rcar/gen4/soc.h (hunk 15 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/soc.h]: fixed

File: drivers/clock_control/clock_control_renesas_cpg_mssr.h (hunk 16 above)
Comment (aaillet) [reply drivers/clock_control/clock_control_renesas_cpg_mssr.h]: fixed

File: soc/arm/renesas_rcar/gen4/pfc_r8a779f0.c (hunk 22 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/pfc_r8a779f0.c]: fixed

File: boards/arm/rcar_spider/rcar_spider_cr52.dts (hunk 23 above)
Comment (aaillet) [reply boards/arm/rcar_spider/rcar_spider_cr52.dts]: why not, as it it well commented

Comment (aaillet) [reply boards/arm/rcar_spider/rcar_spider_cr52.dts]: fixed
removed scif3 from spider board dts
kept it as soc level dtsi and in pinctrl board file
correctly commented soc (r8a779f0) dtsi to show which scif is used by which world

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi (hunk 3 above)
Comment (aaillet) [reply dts/arm/renesas/rcar/gen4/r8a779f0.dtsi]: fixed

File: dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi (hunk 4 above)
Comment (aaillet) [reply dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi]: fixed

File: drivers/pinctrl/pfc_rcar.c (hunk 2 above)
Comment (aaillet) [reply drivers/pinctrl/pfc_rcar.c]: added a protection

File: dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi (hunk 20 above)
Comment (aaillet) [reply dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi]: Same as linux : https://github.com/renesas-rcar/linux-bsp/blob/v5.10.147/rcar-5.2.0.rc6/arch/arm64/boot/dts/renesas/r8a779f0.dtsi#L368

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi (hunk 18 above)
Comment (aaillet) [reply dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:30]: explained in commit msg

File: dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi (hunk 19 above)
Comment (aaillet) [reply dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi]: explained in commit msg
not same domain as control/app domain, internal clock controller domains

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi (hunk 21 above)
Comment (aaillet) [reply dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:21]: epxlained that last four registers were located in control domain of the soc

File: boards/arm/rcar_spider/doc/img/rcar_s4_block_diagram.jpg
Comment (pmarzin) [reply boards/arm/rcar_spider/doc/img/rcar_s4_block_diagram.jpg:1]: ok

File: boards/arm/rcar_spider/rcar_spider_cr52_defconfig (hunk 24 above)
Comment (aaillet) [reply boards/arm/rcar_spider/rcar_spider_cr52_defconfig]: fixed

**Comment by aaillet** [comment]
PR ready for upstream repository !

**Pull request closed without merging**